hawk/lib/hb_report.rb

```diff
@@ -0,0 +1,122 @@
+  # don't use them prior to a generate run, or you'll get the wrong path.
+  # Lastexit is global (this is freaky/dumb - everything should be based
+  # off path, and callers need to be updated to understand this - this
+  # will happen when we allow multiple hb_report runs, as
+  # hb_reports_controller is what cares about lastexit)
+  attr_reader :path, :outfile, :errfile, :lastexit
+
+  # Call with e.g.:
+  #  ("#{Rails.root}/tmp/pids/hb_report", "/tmp/hb_report-hawk").
+  # pidfile will be set based on 'tmpbase', hb_report will be generated
+  # at 'path'.  If path is nil, generate won't work -- make sure you set
+  # it ASAP, before calling generate or relying on outfile or errfile!
+  def initialize(name = nil)
+    tmpbase = Rails.root.join('tmp', 'pids')
+    reports = Rails.root.join('tmp', 'reports')
+
+    tmpbase.mkpath unless tmpbase.directory?
+    reports.mkpath unless reports.directory?
+
+    @pidfile = tmpbase.join("report.pid").to_s
+    @exitfile = tmpbase.join("report.exit").to_s
+    @timefile = tmpbase.join("report.time").to_s
+    if name
+      @path = reports.join(name).to_s
+      @outfile = reports.join("#{name}.stdout").to_s
+      @errfile = reports.join("#{name}.stderr").to_s
+    else
+      @path = nil
+      @outfile = tmpbase.join("report.stdout").to_s
+      @errfile = tmpbase.join("report.stderr").to_s
+    end
+    @lastexit = File.exists?(@exitfile) ? File.new(@exitfile).read.to_i : nil
+  end
+
+  def path=(path)
+    @path = path.to_s
+    @outfile = "#{@path}.stdout"
+    @errfile = "#{@path}.stderr"
+  end
+
+  def running?
+    Util.child_active(@pidfile)
+  end
+
+  # Returns [from_time, to_time], as strings.  Note that to_time might be
+  # an empty string, if no to_time was specified when calling generate.
+  def lasttime
+    File.exists?(@timefile) ? File.new(@timefile).read.split(",", -1) : nil
+  end
+
+  # contents of errfile as array
+  def err_lines
+    err = []
+    begin
+      File.new(@errfile).read.split(/\n/).each do |e|
+        next if e.empty?
+        err << e
+      end if File.exists?(@errfile)
+    rescue ArgumentError => e
+      # This will catch 'invalid byte sequence in UTF-8' (bnc#854060)
+      err << "ArgumentError: #{e.message}"
+    end
+    err
+  end
+
+  # contents of errfile as array, with "INFO" lines stripped (e.g. for
+  # displaying warnings after an otherwise successful run)
+  def err_filtered
+    err_lines.select do |e|
+      !e.match(/( INFO: |(cat|tail): write error)/) && !e.match(/^tar:.*time stamp/)
+    end
+  end
+
+  # Note: This assumes pidfile doesn't exist (will always blow away what's
+  # there), so there's a possibility of a race (or lost hb_report status)
+  # if two clients kick off generation at almost exactly the same time.
+  # from_time and to_time (if specified) are expected to be in a sensible
+  # format (e.g.: iso8601)
+  def generate(from_time, to_time, all_nodes = true)
```

Comment:
Metrics/AbcSize: Assignment Branch Condition size for generate is too high. [30.61/15]